### PR TITLE
make dd.to_csv accept name_function and asterisk in filename for multiple file write support

### PR DESCRIPTION
**Pull request by nirizr**
Signed-off-by: Nir Izraeli <nirizr@gmail.com>

### CODE REVIEW

**Comment by nirizr** [comment]
I made all to_csv tests use dask.dataframe.read_csv instead of pandas.dataframe.read_csv.
Pandas' version didn't support reading multiple files needed for the last test.
This made those tests a bit strange, because unnamed index got a different name ("Unnamed: 0" instead of None).

File: dask/dataframe/io.py [hunk 1]
```diff
@@ -637,23 +637,37 @@ def to_castra(df, fn=None, categories=None, sorted_index_column=None,
...
-                        (df._name, 0), filename, kwargs)
-
-    kwargs2 = kwargs.copy()
-    kwargs2['mode'] = 'a'
-    kwargs2['header'] = False
+                        (df._name, 0), filename.replace('*', name_function(0)), kwargs)
 
     for i in range(1, df.npartitions):
+        filename_i = filename.replace('*', name_function(i))
         dsk[(name, i)] = (_link, (name, i - 1),
```
Comment (dukebody) [inline dask/dataframe/io.py]: Here I don't get why dumping one df to_csv depends on the previous one. This should only happen if a globstring is not provided and the mode is "append". But in the rest of cases the files can be written in parallel. Am I wrong here?

Comment (nirizr) [inline dask/dataframe/io.py]: We aim to keep the order of partitions, and that means one must follow the other, in the same way they are partitioned.

This is a first shot that's only meant to support writing to multiple files. i'm postponing any optimizations here until i'm done with them on `to_hdf`.

If you checkout PR #1293, you'll see a work in progress trying to eliminate `_link`s when they are unnecessary in `to_hdf` but that's not that straightforward

Comment (dukebody) [inline dask/dataframe/io.py]: > We aim to keep the order of partitions, and that means one must follow the other, in the same way they are partitioned.

Yup but I guess that keeping the order of partitions when writing to multiple files consists on naming them using a sorted scheme like `part-001.csv`, `part-002.csv`, etc., where each part file corresponds to a partition, but you can write all these part files at the same time, you don't need the first one to be written before starting with the second one, right?

This is in constrast with the single file mode, where you need the each part to be written (especially the first one, that contains the headers) before being able to write the following one to keep the order.

Comment (nirizr) [inline dask/dataframe/io.py]: This was not part of the goal of this PR as far as i'm concerned, but it was easy enough to add, what do you think?

**Comment by mrocklin** [comment]
Is there a pandas keyword argument to read_csv that would allow us to ignore the index when reading?

**Comment by nirizr** [comment]
if i understand you correctly - yes, `index_col`. but passing it to dask yields an error asking you call `set_index` afterwards.

**Comment by nirizr** [comment]
It wasn't so bad if `dask.dataframe.read_csv` was generating columns named with empty strings instead of "Unnamed: 0", because that's the names you'd have before saving and the names `pandas.read_csv` would give.

File: dask/dataframe/io.py [hunk 2]
```diff
@@ -637,26 +637,54 @@ def to_castra(df, fn=None, categories=None, sorted_index_column=None,
...
+        formatted_names = [name_function(i) for i in range(df.npartitions)]
+        if formatted_names != sorted(formatted_names):
+            warn("In order to preserve order between partitions "
+                 "name_function must preserve the order of its input")
+
+        if filename.count('*') > 1:
+            raise ValueError("A maximum of one asterisk is accepted in filename")
+
+        if 'mode' in kwargs and kwargs['mode'] != 'w':
+            raise ValueError("to_csv does not support writing to multiple files in append mode, "
```
Comment (dukebody) [inline dask/dataframe/io.py]: Hmm I don't know if assuming the user used 'a' mode if she didn't use 'w' mode is the best. Perhaps you can test for kwargs['mode'] == 'a'?

File: dask/dataframe/io.py [hunk 3]
```diff
@@ -637,26 +637,54 @@ def to_castra(df, fn=None, categories=None, sorted_index_column=None,
...
+    if name_function is None:
+        name_function = build_name_function(df.npartitions - 1)
+
+    if '*' in filename:
+        formatted_names = [name_function(i) for i in range(df.npartitions)]
+        if formatted_names != sorted(formatted_names):
+            warn("In order to preserve order between partitions "
+                 "name_function must preserve the order of its input")
+
+        if filename.count('*') > 1:
```
Comment (dukebody) [inline dask/dataframe/io.py]: I'd move this asterisks count test up after `if '*' in filename`, because you don't need to apply the `name_function` if there is more than one asterisk in the file name.

File: dask/dataframe/io.py [hunk 4]
```diff
@@ -637,26 +637,54 @@ def to_castra(df, fn=None, categories=None, sorted_index_column=None,
...
+
+        if 'mode' in kwargs and kwargs['mode'] != 'w':
+            raise ValueError("to_csv does not support writing to multiple files in append mode, "
+                             "please specify mode='w'")
+
+        link = False
+    else:
+        kwargs2['mode'] = 'a'
+        kwargs2['header'] = False
+        link = True
```
Comment (dukebody) [inline dask/dataframe/io.py]: I'd use a more meaningful flag variable name like `multiple_files` instead of `link`.

File: dask/dataframe/tests/test_io.py [hunk 5]
```diff
@@ -757,11 +757,17 @@ def test_to_hdf_multiple_datasets():
         out = pd.read_hdf(os.path.join(dn, 'data_aa.h5'), '/data')
         tm.assert_frame_equal(out, df.iloc[2:])
 
-    # saving to different datasets in multiple files with custom name_function
+    # triggering too many asterisks error
     with tmpdir() as dn:
         with pytest.raises(ValueError):
             fn = os.path.join(dn, 'data_*.h5')
-            a.to_hdf(fn, '/data_*', name_function=lambda i: 'a' * (i +  1))
+            a.to_hdf(fn, '/data_*')
```
Comment (dukebody) [inline dask/dataframe/tests/test_io.py]: Does this belong to this PR??

File: dask/dataframe/tests/test_io.py [hunk 6]
```diff
@@ -820,15 +832,88 @@ def test_to_csv():
         a = dd.from_pandas(df, npartitions)
         with tmpfile('csv') as fn:
             a.to_csv(fn)
-            result = pd.read_csv(fn, index_col=0)
+            result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
```
Comment (dukebody) [inline dask/dataframe/tests/test_io.py]: I'd create multiple test functions, with docstrings. Having all tests in one single function makes it much harder to know what is failing when it does.

File: dask/dataframe/tests/test_io.py [hunk 7]
```diff
@@ -820,15 +832,88 @@ def test_to_csv():
...
+        result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+        result.index.name = None
+        eq(result, df16)
+
+    # test handling existing files when links are optimized out
+    df = pd.DataFrame({'x': ['a', 'b', 'c', 'd'],
+                       'y': [1, 2, 3, 4]}, index=[1., 2., 3., 4.])
+    a = dd.from_pandas(df, 2)
+    with tmpdir() as dn:
+        with pytest.raises(ValueError):
```
Comment (dukebody) [inline dask/dataframe/tests/test_io.py]: Can we move the `with pytest.raises(ValueError)` next to the line where the `ValueError` should be raised? Otherwise it's harder to know which line do we expect to raise the exception.

File: dask/dataframe/tests/test_io.py [hunk 8]
```diff
@@ -820,15 +832,88 @@ def test_to_csv():
...
+            fn = os.path.join(dn, 'data_1.csv')
+            a.to_csv(fn)
+            fn = os.path.join(dn, 'data_*.csv')
+            a.to_csv(fn, mode='a')
+            result = dd.read_csv(fn).set_index('Unnamed: 0').compute()
+            result.index.name = None
+            eq(result, df.append(df))
```
Comment (dukebody) [inline dask/dataframe/tests/test_io.py]: If what we are testing is that an exception is raised, we don't need to further test for equality of any result, right?

File: dask/dataframe/io.py (hunk 2 above)
Comment (nirizr) [inline dask/dataframe/io.py]: Besides 'a', there's also 'r+' to consider. More importantly, 'w' is the default in calls down the line (pd.read_csv).
We can explicitly accept mode defaulting to 'w' as well, but that felt a bit bloat-ish.
I don't particularly care and can add that if you think its better to be explicit here.

File: dask/dataframe/io.py (hunk 3 above)
Comment (nirizr) [inline dask/dataframe/io.py]: done, also moved the mode check.

File: dask/dataframe/tests/test_io.py (hunk 8 above)
Comment (nirizr) [inline dask/dataframe/tests/test_io.py]: Since this error is manually raised, I wanted it there for completeness sake. In case anybody tries supporting that case, he'll have a basic test ready.

File: dask/dataframe/tests/test_io.py (hunk 5 above)
Comment (nirizr) [inline dask/dataframe/tests/test_io.py]: Not really, just saw it from a previous PR of mine and thought it'd be wasteful to change in it's own pr.
It might be a better fit in my `to_hdf` PR, but cba.

File: dask/dataframe/io.py (hunk 4 above)
Comment (nirizr) [inline dask/dataframe/io.py]: this flag is essentially controlling wether we're linknig tasks or not.
In this PR its identical to whether there're multiple files, but in to_hdf it's a bit more complicated.
I wanted to keep the meaning.

File: dask/dataframe/tests/test_io.py (hunk 6 above)
Comment (nirizr) [inline dask/dataframe/tests/test_io.py]: I'm not experiencing that difficulty, and that seems to be the status quo, but there are a lot of tests there, i'll split them to several functions.

File: dask/dataframe/tests/test_io.py (hunk 7 above)
Comment (nirizr) [inline dask/dataframe/tests/test_io.py]: done

**Comment by nirizr** [comment]
Looks like allowing the `index_col` parameter into pd.read_csv worked in replacing `set_index(...)`.
It didn't break any test (except one testing the error itself), but I'm not sure why it was prevented in the first place and if it's really safe to remove.

**Comment by dukebody** [comment]
Can't we use the following?

1) When doing `to_csv`, `index=False`. This will avoid writing a first unnamed column as index.
2) When reading the csv with `read_csv`, if I'm not wrong, the default behaviour is not to use any column as index, but create a new numeric auto-incremental one, right? This is different behavior from `DataFrame.from_csv` that @jreback mentioned, that uses the first column as index by default.

**Comment by nirizr** [comment]
1. I think saving the index by default is good behavior and we should keep it. Especially when saving to multiple files. That also pandas default behavior as you mentioned.
2. Not only `dask.datafame.DataFrame.from_csv` doesn't default to treating the first column as index (the way pandas does), passing that parameter (`index_col`) is forbidden and raises an exception. I'd like to remove that protection, after understanding why it was there in the first place.

**Comment by nirizr** [comment]
Last commit removed the error on `index_col` passed and uses it in my tests.
again, I'd be careful before merging it without understanding why it wasn't allowed.

**Comment by mrocklin** [comment]
Index_col is forbidden in dask.dataframe because dask.dataframe assumes that either the index column is sorted or the index column has no real information (just an integer index.)  We're unable to easily handle an unsorted index, which is quite possible to find in a CSV file.  We ask users to perform a set_index operation explicitly so that they know that this is an expensive operation.

**Comment by nirizr** [comment]
how do you want to continue? 
should I remove the last commit (2596c87) and just leave it with `set_index("Unnamed: 0")`?
do we need to use `set_index` when reading a CSV file saved with dask? perhaps let the user decide?
I don't think I'm capable of making any decision here.

**Comment by mrocklin** [comment]
If it is possible then I think it's best to change tests added in this PR to not write out index columns when not necessary.  Hopefully this simplifies those tests.

I think that we should leave the warning for `index_col` in place

File: dask/dataframe/io.py (hunk 4 above)
Comment (mrocklin) [inline dask/dataframe/io.py]: I suggest that we keep `link` as it is now for consistency and then change it throughout the codebase in a future PR.  

I agree with @dukebody that while the term `link` makes a lot of sense to us as developers it's probably not clear from a naive user perspective.

File: dask/dataframe/tests/test_io.py (hunk 5 above)
Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Why the change?

Comment (nirizr) [inline dask/dataframe/tests/test_io.py]: The test was over complicated. This tests erroring on multiple asterisks, the name function argument was only there because of a copy/paste.

Comment (mrocklin) [inline dask/dataframe/tests/test_io.py]: Ah I see.  That makes sense

File: dask/dataframe/io.py (hunk 2 above)
Comment (nirizr) [inline dask/dataframe/io.py]: @mrocklin Any opinion here?

File: dask/dataframe/io.py (hunk 4 above)
Comment (nirizr) [inline dask/dataframe/io.py]: There's no such code in dask's codebase, only in two of my PRs, i can change both.

File: dask/dataframe/io.py (hunk 2 above)
Comment (mrocklin) [inline dask/dataframe/io.py]: I have no objections to anything that limits scope to the common case.  This appears to limit scope to a simple case that we know we can handle well, so I'm +1.  I especially like that it includes an informative error message.

Obviously it'd be nice to support other modes in the future, but that would involve a lot more work to verify all cases.  Presumably there is higher priority work around.

**Comment by mrocklin** [comment]
This looks good to me.  It appears that the only outstanding comment is around the use of the term `link`.  Do we want to handle that here?

**Comment by nirizr** [comment]
Yes, that's pretty much it. I don't think we should wait for the name, as I'm not sure to_hdf's meaning is that similar anymore.
I do wanna make one last pass. Squash commits and review the final version.I'll try doing that today.

**Comment by nirizr** [comment]
Squashed to a single commit and changed the variable name to `single_file`, I'm good to go.

**Comment by mrocklin** [comment]
Looks great to me.  Merging.

**Comment by mrocklin** [comment]
Thanks @nirizr ! This is in.
